pkg/jobs/jobsprotectedts/jobs_protected_ts_manager.go

```diff
@@ -0,0 +1,225 @@
+		v.ProtectedTimestampRecord = u
+		return v
+	default:
+		panic("not supported")
+	}
+}
+
+func getProtectedTSOnJob(details jobspb.Details) *uuid.UUID {
+	switch v := details.(type) {
+	case jobspb.NewSchemaChangeDetails:
+		return v.ProtectedTimestampRecord
+	case jobspb.SchemaChangeDetails:
+		return v.ProtectedTimestampRecord
+	default:
+		panic("not supported")
+	}
+}
+
+// NewManager creates a new protected timestamp provider
+// for a job.
+func NewManager(
+	db *kv.DB,
+	codec keys.SQLCodec,
+	protectedTSProvider protectedts.Provider,
+	systemConfig config.SystemConfigProvider,
+	jr *jobs.Registry,
+) *Manager {
+	return &Manager{
+		db:                  db,
+		codec:               codec,
+		protectedTSProvider: protectedTSProvider,
+		systemConfig:        systemConfig,
+		jr:                  jr,
+	}
+}
+
+// ProtectBeforeGC adds a protected timestamp record for a historical
+// transaction for a specific table, once a certain percentage of the GC time has
+// elapsed. Returns a Cleaner function to cancel installation or remove the
+// protected  timestamp.
+func (p *Manager) ProtectBeforeGC(
+	ctx context.Context,
+	jobID jobspb.JobID,
+	tableDesc catalog.TableDescriptor,
+	readAsOf hlc.Timestamp,
+) Cleaner {
+	waitGrp := ctxgroup.WithContext(ctx)
+	protectedTSInstallCancel := make(chan struct{})
+	var unprotectCallback Cleaner
+	waitGrp.GoCtx(func(ctx context.Context) error {
+		// If we are starting up the system config can be nil, we are okay letting
+		// the job restart, due to the GC interval and lack of protected timestamp.
+		systemConfig := p.systemConfig.GetSystemConfig()
+		if systemConfig == nil {
+			return nil
+		}
+		// Determine what the GC interval is on the table, which will help us
+		// figure out when to apply a protected timestamp, as a percentage of this
+		// time.
+		zoneCfg, err := systemConfig.GetZoneConfigForObject(p.codec,
+			config.ObjectID(tableDesc.GetID()))
+		if err != nil {
+			return err
+		}
+		waitBeforeProtectedTS := time.Duration(float64(time.Duration(zoneCfg.GC.TTLSeconds)*time.Second) *
+			timedProtectTimeStampGCPct)
+
+		select {
+		case <-time.After(waitBeforeProtectedTS):
+			unprotectCallback, err = p.Protect(ctx, jobID, tableDesc, readAsOf)
+			if err != nil {
+				return err
+			}
+		case <-protectedTSInstallCancel:
+			return nil
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+		return nil
+	})
+	return func(ctx context.Context) error {
+		close(protectedTSInstallCancel)
+		if err := waitGrp.Wait(); err != nil {
+			return err
+		}
+		if unprotectCallback != nil {
+			if err := unprotectCallback(ctx); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+}
+
+// Protect adds a protected timestamp record for a historical
+// transaction for a specific table immediately. Returns a Cleaner
+// function to remove the protected timestamp, if one was installed.
+func (p *Manager) Protect(
+	ctx context.Context,
+	jobID jobspb.JobID,
+	tableDesc catalog.TableDescriptor,
```

Comment:
Side note: i might refactor this function to pass a pts Target instead of a tableDesc, so DR could use this infra in Restore and Import.

Reply:
Definitely, should try and get this adopted in other areas to streamline things, and the descriptor in this case was because it was natural on the schema change side.